### ISSUE TITLE
Use command line parameters in callback

### ISSUE THREAD

**Issue by long-greg-c** [fullphat/snarl#15]
Snarl 3.1 V45.25

I am using a static invocation callback, the callback runs an executable on the local machine and this is working OK.

I now need to add some command line parameters to pass to the executable. 

I am able to pass the callback correctly quoted with the command line parameter. 
See log output:
```
11-Jan-2016 23:08:28	LOG	0	TMainWindow.uCopyData(): dwData=1397640195 hWndFrom=00001EFC
11-Jan-2016 23:08:28	LOG	0	g_DoV42Request(): 'isvisible?app-sig=application/DPANotifier&password=123456&uid=NormalMessage'
11-Jan-2016 23:08:29	LOG	0	TMainWindow.uCopyData(): dwData=1397640195 hWndFrom=00001EFC
11-Jan-2016 23:08:29	LOG	0	g_DoV42Request(): 'notify?app-sig=application/DPANotifier&password=123456&uid=NormalMessage&title=My Title&text=A test message&timeout=0&callback="C:\TestExecutable.exe" -dummyParameter&priority=-1'
11-Jan-2016 23:08:29	LOG	0	g_DoNotify(): looking for uid 'NormalMessage' from app 'application/DPANotifier'...
11-Jan-2016 23:08:29	WARN	0	TApp.Show(): class '' not registered, using default...
11-Jan-2016 23:08:29	LOG	0	TNotificationRoster.Add(): applied presence: 0
11-Jan-2016 23:08:29	LOG	0	TNotificationRoster.Add(): placing at 1606,952-1915,1035
11-Jan-2016 23:08:31	LOG	0	CSnarlWindow.WndProc(): client area invoked
11-Jan-2016 23:08:31	LOG	0	legacy mode: doing invoke..
```

The problem is the executable is not called. If I remove the command line parameter then the executable does get called.

**Comment by fullphat** [comment]
I'm not able to test right now but I'm not sure we support parameters in the callback.  If you've not already tried it, try double quotes as follows:

> ...&callback=""C:\TestExecutable.exe" -dummyParameter"&priority=...


Chris
full phat products
__________________________
http://getsnarl.info
Twitter: @snarlhappens

**Comment by long-greg-c** [comment]
Same result with the double quotes see updated log:
```
13-Jan-2016 23:21:57	LOG	0	TMainWindow.uCopyData(): dwData=1397640195 hWndFrom=00002E7C
13-Jan-2016 23:21:57	LOG	0	g_DoV42Request(): 'isvisible?app-sig=application/DPANotifier&password=123456&uid=NormalMessage'
13-Jan-2016 23:21:57	LOG	0	TMainWindow.uCopyData(): dwData=1397640195 hWndFrom=00002E7C
13-Jan-2016 23:21:57	LOG	0	g_DoV42Request(): 'notify?app-sig=application/DPANotifier&password=123456&uid=NormalMessage&timeout=0&callback=""C:\TestExecutable" -dummParamter==someValue"&priority=-1'
13-Jan-2016 23:21:57	LOG	0	g_DoNotify(): looking for uid 'NormalMessage' from app 'application/DPANotifier'...
13-Jan-2016 23:21:57	WARN	0	TApp.Show(): class '' not registered, using default...
13-Jan-2016 23:21:57	LOG	0	TNotificationRoster.Add(): applied presence: 0
13-Jan-2016 23:21:57	LOG	0	TNotificationRoster.Add(): placing at 1606,952-1915,1035
13-Jan-2016 23:21:59	LOG	0	CSnarlWindow.WndProc(): client area invoked
13-Jan-2016 23:21:59	LOG	0	legacy mode: doing invoke...
```

I now see something else unusual. If the parameter includes an equals sign it gets duplicated in the callback. 

I assume the reason the executable is not called is this line 2044 in mSnarl.bas:
```vbnet
        ' /* treat as URL/file/launchable */
        ShellExecute frmAbout.hWnd, vbNullString, Ack, vbNullString, vbNullString, SW_SHOW
```

We would need to split the Ack string into the executable and the arguments. The arguments would go in the 4th arg of the ShellExecute call.

**Comment by fullphat** [comment]
I agree - apologies.  I'll see if I can fix this in the next 4.0 release.  The only workaround I can think of is to create batch files that pass the parameter, although this might not work depending on what you're doing.

Chris
full phat products
__________________________
http://getsnarl.info
Twitter: @snarlhappens

**Comment by long-greg-c** [comment]
Thanks Chris that would be great to see it added to 4.0. 

Unfortunately the batch file workaround wont work for my setup. 

I'm working on an implementation of Snarl for a business customer that has been using it for a few years with a couple of internal applications and it is a new requirement to pass the parameters in the callback.

As my customer is looking for a quick turn around I am willing to make the change myself but unfortunately i don't have access to some of the dependencies to be able to build the 3.1 release on GitHub, namely the source for melon SOS, mCypher, mWin32Monitors, CZippedContent, CdroppedContentLite, mWin32DragDrop, mArgList and olelib.tlb.

Are you able to provide those?

As it is for a business customer I could also pay for your time if you were to make the changes...

**Comment by fullphat** [comment]
Much as I'd love the funds :) at the moment I can't commit the time in afraid.  I'll get you those missing source files ASAP however.

Chris
full phat products
__________________________
http://getsnarl.info
Twitter: @snarlhappens

**Comment by fullphat** [comment]
Here you go:

http://sourceforge.net/projects/snarlwin/files/Snarl/Source%20Code/

"missing-files.zip"

Chris
full phat products
__________________________
http://getsnarl.info
Twitter: @snarlhappens

**Comment by long-greg-c** [comment]
Thanks Chris.

Can you provide these three files too?

```
Class=CDropContentLite; ..\..\..\..\melon\trunk\src\extra\CDropContentLite.cls
Module=SOS_i; ..\..\..\..\melon\trunk\src\sos2\current\include\SOS_i.bas
Class=SOS; ..\..\..\..\melon\trunk\src\sos2\current\include\SOS.cls
```

**Comment by fullphat** [comment]
Done - sorry!

Chris
full phat products
__________________________
http://getsnarl.info
Twitter: @snarlhappens

**Comment by long-greg-c** [comment]
Thanks Chris.
Sorry to say but I had some other issues preventing a compile. The following functions are missing from the source code.
```
prefskit_NewSlider
g_CreateDirectory
prefskit_SetItemObject
g_Capitalise
prefskit_NewImage
prefskit_FindInList
g_PercentToAlpha
g_RemoveLineFeeds
```

Any chance you could provide those?

cheers

**Comment by fullphat** [comment]
Sorry for the delay!  We're back and were up and running with R5 using C#/.net 4.0 :)  I'll get this fixed - how about using a comma in the callback to separate out args?  Example: ...callback=some.exe,args go:here...

**Comment by long-greg-c** [comment]
I think the arguments for the callback should be in a separate option. There are many ways arguments might need to be passed to an executable.
```Example: callback="some executable with spaces.exe" callback_args="-a arg1 anotherType -arg3 arg3 value"```

**Comment by fullphat** [comment]
Yes, makes sense - and nicely backwards compatible as well.  Will add this to R5.0!
